store/sqlstore/channel_store.go

```diff
@@ -757,14 +955,44 @@ func (s SqlChannelStore) saveMemberT(transaction *gorp.Transaction, member *mode
 		return result
 	}
 
-	if err := transaction.Insert(member); err != nil {
+	dbMember := NewChannelMemberFromModel(member)
+
+	if err := transaction.Insert(dbMember); err != nil {
 		if IsUniqueConstraintError(err, []string{"ChannelId", "channelmembers_pkey"}) {
 			result.Err = model.NewAppError("SqlChannelStore.SaveMember", "store.sql_channel.save_member.exists.app_error", nil, "channel_id="+member.ChannelId+", user_id="+member.UserId+", "+err.Error(), http.StatusBadRequest)
 		} else {
 			result.Err = model.NewAppError("SqlChannelStore.SaveMember", "store.sql_channel.save_member.save.app_error", nil, "channel_id="+member.ChannelId+", user_id="+member.UserId+", "+err.Error(), http.StatusInternalServerError)
 		}
 	} else {
-		result.Data = member
+		var retrievedMember channelMemberWithSchemeRoles
+		if err := transaction.SelectOne(&retrievedMember, `
+			SELECT
```

Comment:
I have to look into more detail in this query. My concerns are, performance (I'm almost sure the performance must be good, but I prefer to look at the SQL Plan to verify it), the absence of some results (With LEFT JOIN, you discard any NULL value at the right, so if you have not ChannelScheme or TeamScheme, for that membership could be filtered, probably it can't happen in a consistent database), the other point, are duplications (And I think is something that can't happen, but I prefer to take a deeper look to it.

Anyway LGTM at first sight.

Reply:
Agree that we'll have to test on performance. If it's bad, JW suggested some other ways to approach that I can try.

I think the JOIN is correct - it will return the left side regardless of if there's a matching right side, and the join is on the right side's primary key so it can only return 0 or 1 matching rows on the right.